### ISSUE TITLE
Implement flake8-import-restrictions

### ISSUE THREAD

**Issue by justinchuby** [astral-sh/ruff#5841]
When following the Google style guide and to avoid cyclic imports, we would really like to see the rule from https://github.com/atollk/flake8-import-restrictions#imr241

(import modules only and not the functions and classes from the modules)

> IMR241
> When using the from syntax, only submodules are imported, not module elements.

```
# Bad
from os.path import join

# Good
from os import path
```

**Comment by justinchuby** [comment]
So it’s a duplicate of https://github.com/astral-sh/ruff/issues/3045, but I found flake8-import-restrictions which has a good collection of related rules.

**Comment by edgarrmondragon** [comment]
It's not exactly equivalent, but the [`banned-from`](https://beta.ruff.rs/docs/settings/#flake8-import-conventions-banned-from) setting can get a similar violation with this config:

```toml
# ruff.toml
[flake8-import-conventions]
banned-from = ["os.path"]
```

```console
$ ruff check --select ICN myfile.py
myfile.py:2:1: ICN003 Members of `os.path` should not be imported explicitly
Found 1 error.
```

**Comment by justinchuby** [comment]
True but it is impossible to configure the list for all possible packages. It would be good to recognize an imported entity is not a module and error on that.

**Comment by charliermarsh** [comment]
We would need to integrate our Python module resolver to support this (doable).

**Comment by QuentinSoubeyranAqemia** [comment]
I would also be interested in a rule to disallow importing members, and only allow importing module (or packages, but they are module too technically speaking), per the Google Code Style.
Having a config option for exceptions (like the `typing`, `typing_extensions` and `collections.abc` defined in the google style guide) would be most useful too.

I'm curious what is this "Python module resolver"? Is this an upcoming feature? Some internal that would need to be exposed to rules? Something devs are thinking of doing?

**Comment by segoon** [comment]
Hello,

I might miss something, but AFACS the technique used in https://github.com/atollk/flake8-import-restrictions/blob/master/flake8_import_restrictions/imports_submodule.py#L35 is vulnerable to arbitrary code execution due to the loading of python module from the working directory. I'd not run such a check on my system account with a random foreign code.

I've tried to implement the same part of Google style guide in https://github.com/astral-sh/ruff/pull/22847 another way: instead of loading the module and explicit check of imported symbol type, we use a heuristics that "if the symbol is used as a non-module, then it must be a type/function, otherwise assume it is indeed a module". The check is autonomous, an isolated module code is enough, no external module is required for validation.

**Comment by segoon** [comment]
@justinchuby @chirizxc @ntBre Hello, any update please?

**Comment by chirizxc** [comment]
Actually, I think it's strange, even if they do it, this rule should remain in preview forever, like the `E...` rules.

**Comment by ntBre** [comment]
We'll need type inference, or at least the multi-file analysis and module resolver that are being built as part of ty to implement this in Ruff.

**Comment by segoon** [comment]
Can you please clarify what's wrong with implementation in https://github.com/astral-sh/ruff/pull/22847? What specific cases does it miss? I agree the check might miss something, but as for my codebase it just works with no false positives.

**Comment by ntBre** [comment]
The issue isn't with the implementation, it's that the rule itself (this issue) is still labeled https://github.com/astral-sh/ruff/labels/needs-decision, which indicates to me that there's not strong consensus that we want to include this rule. I'm also worried about false positives without type inference since we have no fully reliable way to verify the type of an imported object. I will approve the workflow runs on your PR so that we can see the ecosystem report. I see that you have some more clever heuristics than simply checking the case of names, but I'm still not sure we want to add this rule at the moment either way.

**Comment by segoon** [comment]
Hi, I've pushed a test fix. Any more suspicions, problems with the PR?

**Comment by segoon** [comment]
@ntBre

**Comment by chirizxc** [comment]
As @ntBre noted in https://github.com/astral-sh/ruff/issues/5841#issuecomment-3921107775, it is still labeled [needs-decision](https://github.com/astral-sh/ruff/labels/needs-decision)

I think we should hear a few opinions for and against this rule (even this issue shows how unpopular this rule is, unlike others).

**Comment by segoon** [comment]
Anything else can I provide to speed up this process?

**Comment by seminumber** [comment]
Our team do need this rule as we adopt google python style guide. I have limited expertise but for about false-positives, we'd be happy to have features early and would be happier if we got some features like list of imports to be regarded as exceptions for this rule. IMHO, other tools like Pycharm make use of the kind of heuristic for auto-import; most packages are hueristic-friendly while others may not -- and those who use these tools are well aware that they are not linter-friendly -- scikit-learn for instance, and if there are false positives on these packages, I think people will understand because they are so with other tools. Having that in mind, maybe you can support typestubs for this feature as well to make less false positives in some cases.

EDIT: Duplicates #3045 ?

**Comment by serjflint** [comment]
Hi,
Can this issue be reconsidered with introduction of https://github.com/astral-sh/ruff/issues/1774 or is any heuristic is blocked by decision on https://github.com/astral-sh/ruff/issues/7447?
